### ISSUE TITLE
support turborepo when using dependenciesMeta.*. injected (make injected deps re-sync)

### ISSUE THREAD

**Issue by NullVoxPopuli** [pnpm/pnpm#6088]
- [x] I'd be willing to implement this feature ([contributing guide](https://github.com/pnpm/pnpm/blob/main/CONTRIBUTING.md))

I might be willing to implement if i know what the plan is.

Very similar to: 
- https://github.com/pnpm/pnpm/issues/4407

Related to:
- https://github.com/pnpm/pnpm/issues/4965
- https://github.com/vercel/turbo/issues/2306
- https://github.com/CrowdStrike/ember-headless-form/pull/49
  - https://github.com/CrowdStrike/ember-headless-form/pull/45
  - https://github.com/CrowdStrike/ember-headless-form/pull/52
- https://github.com/CrowdStrike/ember-headless-table/pull/44

**Describe the user story**

When using dependenciesMeta.*.injected, and turborepo, the built packages are not sync'd.

**Describe the solution you'd like**

I would like the dependencies to be sync'd when using dependenciesMeta.*. injected.

**Describe the drawbacks of your solution**

I can't think of any. Afaik, this is the most correct way to have dependencies?

**Describe alternatives you've considered**

I haven't thought of any

**Comment by NullVoxPopuli** [comment]
@simonihmig came up with a way to make this work over at: https://github.com/CrowdStrike/ember-headless-form/pull/52

It requires this util: https://github.com/CrowdStrike/ember-headless-form/pull/52/files#diff-cba6a6cd1877ea907fa4d3c06bdd8548011bc52b603638c9e8334870e53aca39 which manually re-hard-links the dependencies.

does it make sense for to add this util to pnpm?

It would at least give us a quicker command than `pnpm i -f` (and `pnpm i -f` can sometimes break dev servers)

**Comment by zkochan** [comment]
Doesn't it re-sync on `pnpm install`? I guess it should be fixed in that case, I think it should definitely do re-sync on `pnpm install`.

**Comment by NullVoxPopuli** [comment]
@zkochan it turns out that we can't use `pnpm i` in place of @simonihmig's script anyway, because we need to be able to run the script in parallel, whereas `pnpm i` frequently errors when multiple processes doing `pnpm i` run at the same time. 

So, I would propose we add a `pnpm sync` command that does what @simonihmig's script does, within a workspace, ensures that any injected dependencies are synced -- this would allow turboroepo to be used with pnpm and injected dependencies.

I'll start a PR

**Comment by LincWong** [comment]
i really hope this feature release soon...:(

**Comment by runspired** [comment]
yeah this would be very nice :)

**Comment by LincWong** [comment]
你好，我已经收到你的邮件了。

**Comment by nxmad** [comment]
That would be a very nice feature, however I think it's more CI-related as it does not solve the problem in dev mode.
Here is our scenario (which is pretty common I believe):
- pnpm install
- build "internal" packages
- build application and run dev server

Dev server will just ignore built code of internal packages as it was not hardlinked during pnpm install. So we will need to run pnpm sync every time they changed.
I guess in addition to pnpm sync command we could add `--ignore-injected` flag for pnpm install in dev environment, so pnpm will rollback it's behaviour to default symlinking.

**Comment by runspired** [comment]
@nxmad it actually is exactly for solving the issue in dev mode! It would allow the file watcher of something like Vite or Turborepo or Turbopack to trigger a change notification for the hardlink when the dep changes.

**Comment by NullVoxPopuli** [comment]
> i really hope this feature release soon...:(

@LincWong, I started a PR here: https://github.com/pnpm/pnpm/pull/6135 (not ready for review), and once I finish, I can also provide an example of how to hook up turborepo to handle "everything" automatically

**Comment by simonihmig** [comment]
> I can also provide an example of how to hook up turborepo to handle "everything" automatically

@NullVoxPopuli does "everything" also include re-syncing after a dependency was re-built in watch-mode? If so, please share what you have in mind! 😀

For me indeed dev mode is still an unsolved problem: I have to explicitly run the sync script mentioned above from inside of the app whenever `rollup -w` updates the `/dist` output of a workspace dependency. 😕

That watch mode might actually be best implemented as part of the task runner, as it knows already the dependency graph:
* there is some discussion around a [watch mode for turborepo](https://github.com/vercel/turbo/issues/986), which seems to have positive feedback, just not a priority atm
* `nx` seems to already [support watching](https://nx.dev/recipes/other/workspace-watching). Haven't tested this yet though...

**Comment by gron1gh1** [comment]
Hello, I'm very much in favor of this issue. I suffered a lot from these problems, and I made a library as a temporary measure.

https://github.com/gron1gh1/lopm
**It provides a command to analyze local packages and hardlink them to node_modules, and supports watch mode.**

Please refer to it if you need it, and I think it will be unnecessary if this issue is provided normally.
If there is a problem, I will delete it. Thank you.

**Comment by NullVoxPopuli** [comment]
Added a README to this project: https://github.com/NullVoxPopuli/pnpm-sync-dependencies-meta-injected

**Comment by LincWong** [comment]
你好，我已经收到你的邮件了。

**Comment by ef4** [comment]
This problem is not unique to turborebo. Even normal git operations will break the hard linking.

For example, `git stash` a change to a file and then `git pop` it back out. You'll find you lost the hardlink because git did an `unlink` and `create` instead of `open` and `write`.

There's just a lot of possible things that can cause desynchronization. A watcher is necessary to really use per-file hard linking.

**Comment by LincWong** [comment]
你好，我已经收到你的邮件了。

**Comment by NullVoxPopuli** [comment]
> A watcher is necessary to really use per-file hard linking

idk how you feel about _temporary tools_ to get around problems, or to assess DX, but https://github.com/NullVoxPopuli/pnpm-sync-dependencies-meta-injected has a watch mode,
```json
"start": "concurrently 'your dev command' 'pnpm _syncPnpm --watch'",
"_syncPnpm": "pnpm sync-dependencies-meta-injected"
```

**Comment by runspired** [comment]
this looks to be a duplicate of #4407

**Comment by LincWong** [comment]
你好，我已经收到你的邮件了。

**Comment by zkochan** [comment]
Closing as a duplicate of #4407 (2022-03-04), which is the older report of the same thing.

Same request: an injected workspace dependency should be copied again after the source project is rebuilt.

Follow #4407 for this. If you think the two differ in a way that matters, say so here and we will reopen.

---
Written by an agent (Claude Code, claude-opus-5).
